api/v1alpha1/internalrequest_types.go

```diff
@@ -57,9 +90,78 @@ type InternalRequest struct {
 	Status InternalRequestStatus `json:"status,omitempty"`
 }
 
+// HasCompleted checks whether the InternalRequest has been completed.
+func (ir *InternalRequest) HasCompleted() bool {
+	condition := meta.FindStatusCondition(ir.Status.Conditions, InternalRequestSucceededConditionType)
+	return condition != nil && condition.Status != metav1.ConditionUnknown && ir.Status.CompletionTime != nil
+}
+
+// HasFailed checks whether the InternalRequest has failed.
+func (ir *InternalRequest) HasFailed() bool {
+	return meta.IsStatusConditionFalse(ir.Status.Conditions, InternalRequestSucceededConditionType)
+}
+
+// HasStarted checks whether the InternalRequest has started.
+func (ir *InternalRequest) HasStarted() bool {
+	condition := meta.FindStatusCondition(ir.Status.Conditions, InternalRequestSucceededConditionType)
+	return condition != nil && ir.Status.StartTime != nil
+}
+
+// HasSucceeded checks whether the InternalRequest has succeeded.
+func (ir *InternalRequest) HasSucceeded() bool {
+	return meta.IsStatusConditionTrue(ir.Status.Conditions, InternalRequestSucceededConditionType)
+}
+
+// MarkFailed registers the completion time and changes the Succeeded condition to False with the provided message.
+func (ir *InternalRequest) MarkFailed(message string) {
+	if ir.HasCompleted() {
+		return
+	}
+
+	ir.Status.CompletionTime = &metav1.Time{Time: time.Now()}
+	ir.setStatusConditionWithMessage(InternalRequestSucceededConditionType, metav1.ConditionFalse, InternalRequestFailed, message)
+}
+
+// MarkRunning registers the start time and changes the Succeeded condition to Unknown.
+func (ir *InternalRequest) MarkRunning() {
+	if ir.HasStarted() {
+		return
+	}
+
+	ir.Status.StartTime = &metav1.Time{Time: time.Now()}
+	ir.setStatusCondition(InternalRequestSucceededConditionType, metav1.ConditionUnknown, InternalRequestRunning)
+}
+
+// MarkSucceeded registers the completion time and changes the Succeeded condition to True.
+func (ir *InternalRequest) MarkSucceeded() {
+	if ir.HasCompleted() {
+		return
+	}
+
+	ir.Status.CompletionTime = &metav1.Time{Time: time.Now()}
+	ir.setStatusCondition(InternalRequestSucceededConditionType, metav1.ConditionTrue, InternalRequestSucceeded)
+}
+
+// SetCondition creates a new condition with the given InternalRequestSucceededConditionType, status and reason. Then, it sets this new condition,
+// unsetting previous conditions with the same type as necessary.
+func (ir *InternalRequest) setStatusCondition(conditionType string, status metav1.ConditionStatus, reason HandlingReason) {
```

Comment:
Shouldn't the `conditionType`'s type actually be `InternalRequestSucceededConditionType` (even if it's a string underneath)? I'm a total Go noob, so just curious.

This would apply to the `setStatusConditionWithMessage` function as well.

Reply:
In the future we could have other condition types, so this function is more generic.

Reply:
OK, good.